rslib/src/import_export/text/import.rs

```diff
@@ -290,19 +290,23 @@ impl<'a> Context<'a> {
 
     fn import_note(&mut self, ctx: NoteContext, log: &mut NoteLog) -> Result<()> {
         match self.dupe_resolution {
-            _ if ctx.dupes.is_empty() => self.add_note(ctx, log, false)?,
-            DupeResolution::Add => self.add_note(ctx, log, true)?,
+            _ if ctx.dupes.is_empty() => self.add_note(ctx, log)?,
+            DupeResolution::Add => self.add_note(ctx, log)?,
             DupeResolution::Update => self.update_with_note(ctx, log)?,
             DupeResolution::Ignore => log.first_field_match.push(ctx.note.into_log_note()),
         }
         Ok(())
     }
 
-    fn add_note(&mut self, ctx: NoteContext, log: &mut NoteLog, dupe: bool) -> Result<()> {
+    fn add_note(&mut self, ctx: NoteContext, log: &mut NoteLog) -> Result<()> {
         if !ctx.note.first_field_is_unempty() {
             log.empty_first_field.push(ctx.note.into_log_note());
             return Ok(());
         }
+        if ctx.dupes.iter().any(|d| d.note.guid == ctx.note.guid) {
```

Comment:
We simply walk all duplicates, so I'd say it's O(n).
But in fact, `ctx.dupes.get(0).map_or(false, |d| d.note.guid == ctx.note.guid)` should be enough, because if the duplicate was matched by GUID there can only be one.

I'm rethinking the whole GUID-case, anyway. Currently, if the GUID doesn't match, we look for matches by first field. I'm not sure if this makes much sense. Also, we must allow an empty first field for GUID-matches, but disallow duplication. So it might make sense to separate the two cases (GUID vs. no GUID / first-field matching) more strictly.

Reply:
add_note() is O(n), but we call it for each note in the import, so I think the worst case is O(n^2). If we need to do this comparison, we should probably be using a HashSet instead.

Agreed that we should be using GUID or first field, not both, so a separate code path may make more sense.

Reply:
Sorry, nevermind - I mistakenly assumed this was in the context of the whole import, not a single ForeignNote.